hal/src/thumbv7em/icm.rs

```diff
@@ -0,0 +1,1640 @@
+    }
+
+    /// Region Digest Mismatches!( interrupt status
+    #[inline]
+    pub fn get_rdm_int(&self) -> bool {
+        matches!(self.isr().read().rdm().bits() & self.mask(), 1)
+    }
+
+    /// Region Hash Completed interrupt status
+    #[inline]
+    pub fn get_rhc_int(&self) -> bool {
+        matches!(self.isr().read().rhc().bits() & self.mask(), 1)
+    }
+
+    /// When reading the interrupt (ISR) register, it is cleared
+    ///
+    /// This might result in unexpected results for example
+    /// when sequentially trying to determine which interrupt triggered.
+    ///
+    /// This is an alternative, return all the data from the register
+    /// and parse later with the designated `get_[name]_int` functions.
+    #[inline]
+    pub fn get_interrupt_status(&mut self) -> RegionInterrupt<I> {
+        let interrupt = Interrupt(self.isr().read().bits());
+        RegionInterrupt {
+            region: PhantomData,
+            interrupt,
+        }
+    }
+}
+
+/// ICM Peripheral
+///
+/// Encapsulates the PAC which acts as a token
+/// and provides an interface to the ICM hardware
+pub struct Icm {
+    /// ICM pac register providing hardware access
+    icm: crate::pac::ICM,
+}
+
+impl Icm {
+    /// Create the interface for the ICM peripheral
+    ///
+    /// Don't forget to enable the `APB` bus for ICM
+    ///
+    /// `AHB` bus is on by default at reset
+    ///
+    /// Clock::v1
+    /// `mclk.apbcmask.modify(|_, w| w.icm_().set_bit());`
+    ///
+    /// Clock::v2
+    /// `tokens.apbs.icm.enable();`
+    #[inline]
+    pub fn new(icm: crate::pac::ICM) -> Self {
+        Self { icm }
+    }
+
+    // Register Interface
+
+    /// Integrity Check Module
+    #[inline]
+    fn icm(&self) -> &RegisterBlock {
+        &self.icm
+    }
+
+    /// Configuration
+    #[inline]
+    fn cfg(&self) -> &CFG {
+        &self.icm().cfg
+    }
+
+    /// Control
+    #[inline]
+    fn ctrl(&self) -> &CTRL {
+        &self.icm().ctrl
+    }
+
+    /// Region Descriptor Area Start Address
+    #[inline]
+    fn dscr(&self) -> &DSCR {
+        &self.icm().dscr
+    }
+
+    /// Region Hash Area Start Address
+    #[inline]
+    fn hash(&self) -> &HASH {
+        &self.icm().hash
+    }
+
+    /// Interrupt Disable
+    #[inline]
+    fn idr(&self) -> &IDR {
+        &self.icm().idr
+    }
+
+    /// Interrupt Enable
+    #[inline]
+    fn ier(&self) -> &IER {
+        &self.icm().ier
+    }
+
+    /// Interrupt Mask
+    #[inline]
+    fn imr(&self) -> &IMR {
+        &self.icm().imr
+    }
+
+    /// Interrupt Status
+    #[inline]
+    fn isr(&self) -> &ISR {
+        &self.icm().isr
+    }
+
+    /// Status
+    #[inline]
+    fn sr(&self) -> &SR {
+        &self.icm().sr
+    }
+
+    /// Undefined Access Status
+    #[inline]
+    fn uasr(&self) -> &UASR {
+        &self.icm().uasr
+    }
+
+    // User interface for ICM
+
+    /// Enable the ICM peripheral
+    #[inline]
+    pub fn enable(&mut self) {
```

Comment:
I don't quite understand this peripheral, but it seems like there are some things you can/can't do before/after enabling it. Would it be worth adding type safety, like we have for `v2::spi`, `v2::uart` and `clock::v2`? I'm not sure if it would make sense or how difficult it would be.

Reply:
I did have a go at getting some basic type safety, but it became rather cumbersome to use as a RTIC resource, which most likely will be my use case. It is quite possible I got it all wrong, but I get stuck on either requiring dynamic Trait objects, which the compiler isn't happy about since now Icm with the `Is` trait refer to `Self`, or just being unable to lock the RTIC resource and modify it in place since I cannot move out of a reference.

My experimentation can be found here: [HAL icm-typestate][haltypestate] and [demo icm-typestate][demotypestate]

[haltypestate]: https://github.com/vcchtjader/atsamd/tree/icm-typestate
[demotypestate]: https://github.com/vcchtjader/atsamd-demo/tree/icm-typestate

For the bigger picture, looking at the [atmel-software-demo][atmeldemo] they use it as a proper "enable" after configuring the ICM. When ICM has completed its work, they turn it off, then software reset (`swrst()`) and then configure ICM anew before enabling.

Turning it off explicitly is interesting since `swrst()` clears the `enable` bit, maybe they just wanted to be extra clear.

Some observations after testing:

If one set up the ICM for memory monitoring and set the `wrap` bit, causing it to loop back to the start through memory and monitor it continuously then it is easy to observe the `enable` bit, since it is continuously operating.

On the other hand, if there is no wrap bit set, then the `enable` bit acts as a "run once" push-button. Observable not through the `enable` bit (my timing and GDB overhead and all that makes it hard I imagine), but rather through the continuous stream of interrupts if you set the `enable` bit in the interrupt handler, causing yet another interrupt, creating an infinite loop.

With this rather loose usage of the enable bit and the general complexity of the peripheral I think it might be quite tricky to achieve something fully safe while also having the full flexibility if taking the typestate path, especially since it is not one of those "everybody will need this" peripherals. However, if that becomes the case we could always revisit this :smile: 

Thoughts?

[atmeldemo]: https://github.com/atmelcorp/atmel-software-package/blob/master/examples/crypto_icm/main.c